diff --git a/src/jconf.h b/src/jconf.h
index 7504cbb..8cf7cce 100644
--- a/src/jconf.h
+++ b/src/jconf.h
@@ -26,7 +26,7 @@
 #define MAX_CONF_SIZE 16 * 1024
 #define MAX_DNS_NUM 4
 #define MAX_CONNECT_TIMEOUT 10
-#define MAX_UDP_TIMEOUT 120
+#define MAX_UDP_TIMEOUT 10
 
 typedef struct {
     char *host;
diff --git a/src/netutils.c b/src/netutils.c
index ce427b4..17d13c9 100644
--- a/src/netutils.c
+++ b/src/netutils.c
@@ -80,9 +80,9 @@ size_t get_sockaddr(char *host, char *port, struct sockaddr_storage *storage, in
         hints.ai_family = AF_UNSPEC;     /* Return IPv4 and IPv6 choices */
         hints.ai_socktype = SOCK_STREAM; /* We want a TCP socket */
 
-        int err;
+        int err, i;
 
-        for (int i = 1; i < 8; i++) {
+        for (i = 1; i < 8; i++) {
             err = getaddrinfo(host, port, &hints, &result);
             if (!block || !err) {
                 break;
