diff --git a/m4/ax_pthread.m4 b/m4/ax_pthread.m4
index 0b2718f..83909f1 100644
--- a/m4/ax_pthread.m4
+++ b/m4/ax_pthread.m4
@@ -182,6 +182,11 @@ for flag in $ax_pthread_flags; do
                 AC_MSG_CHECKING([whether pthreads work without any flags])
                 ;;
 
+                -pthread)
+                AC_MSG_CHECKING([whether pthreads work with $flag])
+                CFLAGS="-lpthread"
+                ;;
+
                 -*)
                 AC_MSG_CHECKING([whether pthreads work with $flag])
                 PTHREAD_CFLAGS="$flag"
